### ISSUE TITLE
[Bug] Operator kills Galera bootstrap pod even if it is not fully initialized

### ISSUE THREAD

**Issue by djakobczak1** [mariadb-operator/mariadb-operator#421]
**Documentation**
- [x] I acknowledge that I have read the relevant [documentation](https://github.com/mariadb-operator/mariadb-operator/tree/main/docs).

**Describe the bug**

In some environments MariaDB fails to bootstrap with following errors:
```
2024-02-27 15:09:44 2 [Note] WSREP: Synchronized with group, ready for connections
2024-02-27 15:09:44 2 [Note] WSREP: wsrep_notify_cmd is not defined, skipping notification.
2024-02-27 15:09:53 7 [Warning] Access denied for user 'root'@'localhost' (using password: YES)
2024-02-27 15:09:53 8 [Warning] Access denied for user 'root'@'localhost' (using password: YES)
....
```

The access denied logs are from failing liveness probe:
```
  Warning  Unhealthy  2m55s (x5 over 4m55s)   kubelet            Liveness probe failed: HTTP probe failed with statuscode: 500
```

It is caused because it may happen that operator deletes mariadb bootstrap pod before entrypoint initialization completes.
```
$ k logs  mariadb-galera-0  -f
Defaulted container "mariadb" out of: mariadb, agent, init (init)
2024-02-27 15:08:39+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.6.16.11+maria~ubu2004 started.
2024-02-27 15:08:39+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
2024-02-27 15:08:39+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.6.16.11+maria~ubu2004 started.
2024-02-27 15:08:39+00:00 [Note] [Entrypoint]: Initializing database files
# killed
```
Pod is killed after graceful timeout, which by default is set to 30s, which may not be enough to finish initialization process.

**Expected behaviour**

MariaDB installation completes without error.

**W/A**

Increase terminationGracePeriodSeconds of the mariadb pod, which currently is not possible (could be added to MariaDB.Spec and used during mariadb pod template build).

**Steps to reproduce the bug**
Reproduced locally with KIND cluster.

1. Prepare mariadb CR with cpu resource limits
```
  resources:
    limits:
      cpu: 15m
```
2. Install MariaDB CR

**Debug information**

Relevant code snippet:
``` go
# pkg/controller/galera/controller.go
func (r GaleraReconciler) initBootstrapPod(ctx context.Context, mariadb *mariadbv1alpha1.MariaDB, logger logr.Logger) (ctrl.Result, error) {
...
	logger.V(1).Info("Restarting bootstrap Pod")
	
	if err := r.Delete(ctx, &bootstrapPod); err != nil {
		return ctrl.Result{}, fmt.Errorf("error restarting bootstrap Pod: %v", err)
	}
```

**Environment details**:
- Kubernetes version: v1.24.4, v1.28.0
- Kubernetes distribution: WRCP, KIND
- mariadb-operator version: 0.0.26 (d85a5be1d4624c39a7501ea24eff8e66c8119525) - this version is not released yet
- Install method: helm
- Install flavor: custom

**Comment by mmontes11** [comment]
Hey @djakobczak1 ! Thanks for reporting this and also for the detailed analysis.

Currently, the operator waits until the first `Pod` initializeds the `/var/lib/mysql` directory, but it might not be enough. What do you think it would be the best criteria to consider a MariaDB `Pod` initialized? I'm thinking that we could instead perform a `SELECT 1;` with the root credentials to determine this. What are your thoughts about this?

> Pod is killed after graceful timeout, which by default is set to 30s

I believe the `Pod` is being killed by the kubelet, as the liveness probe is failing. During the cluster bootstrap, the operator doesn't kill the `Pod` until is marked as initialized. This timeouts only apply to the recovery process:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/API_REFERENCE.md#galerarecovery

I'm also thinking about provisioning the first PVC and using a Kubernetes `Job` with the MariaDB image to perform the initialization, the same way the first `Pod` currently. This might require extra adaptations to integrate with the entrypoint, so probably something to tackle in further releases.

Thanks again!

**Comment by djakobczak1** [comment]
Hi @mmontes11,

> I believe the Pod is being killed by the kubelet, as the liveness probe is failing. During the cluster bootstrap, the operator doesn't kill the Pod until the agent reports that it is initialized.

There are 2 situation when bootstrap pod can be killed, one is when operator gets information from agent that initialization is completed, the second one is when kubelet kills the pod due to specified number of failed liveness probes. I was referring to the first situation.
Maybe I was not precise enough or I did not get it right, but I thought that following actions happened in described problem:
1. Bootstrap server starts
2. Entrypoint of bootstrap server starts temporary server
3. /var/lib/mysql is initialized
4. Agent side-pod reports that DB is initialized (files were created in /var/lib/mysql, but it does mean that all initialization steps were finished)
5. Operator requests bootstrap pod deletion <- here bootstrap server has by default 30s, which is time for graceful shutdown, to complete initialization, otherwise step 7 will happen
6. Bootstrap server is restarted
7.1 if server was not fully initialized liveness probe fails (loop)
7.2. Bootstrap pod enters CrashLoopBack

Regarding proposed solution (i.e. making sure that we can login to db with changed credentials before restarting the pod), I think it should be fine, because login with changed root password will be possible after restart of temporary server, so user custom scripts placed in /docker-entrypoint-initdb.d/ should not be affected as well.

> One thing we could do is returning 200 OK in the agent /liveness endpoint before the first Pod has been initialized. This way the kubelet will not kill the bootstrap Pod.

Maybe it is worth to think about `startupProbe`? It sounds like `startupProbe` should be used for this kind of situations, when we are waiting for initialization (and with it we can still allow user to configure it, so depending on environment user could adjust for how long it is acceptable to bootstrap the cluster).

Using `Job` has some other advantages e.g. it will help with db migration, which should be done before any other server join the galera cluster (to limit the time needed for certification process).

**Comment by mmontes11** [comment]
Hey @djakobczak1 ! 

You get the process exactly right, again great analysis!

> Maybe it is worth to think about startupProbe? 

This is a good one. It doesn't make sense to start liveness and readiness probes if there initialization hasn't finished. Much better and less risky than adding a condition to the liveness probe.

> Using Job has some other advantages

Yes, apart from what you just mentioned, the advantages I see is that once the init `Job` is finished, we can just create the `StatefulSet` without having to do any sort of coordination between controller, init and agent nor restarting `Pods` during the bootstrap. Which implies that it will be easier to maintain and reduces the risk of race conditions during the bootstrap.

However I also see downsides:
- We need to pre-provision the first PVC to mount it in the `Job`. Currently we rely on the `StatefulSet` for doing this. 
- The `Job` will have a custom entrypoint that, using the [mariadb](https://github.com/MariaDB/mariadb-docker) image, it will:
   - Source the default entrypoint
   - Call the `docker_setup_env` and `docker_mariadb_init` functions
- These functions are only available in certain image versions: the ones available in this [repo](https://github.com/MariaDB/mariadb-docker). This also includes the current enterprise images, they have the same entrypoint.

>  Entrypoint script for MariaDB Server 1:10.6.1

10.6 should be compatible I believe. Is this version the one you are intending to use?

In conclusion, I think I will give the init `Job` a try. If I find any other limitation along the way, checking the root credentials in the `startupProbe`, as you suggested, should also do it.

**Comment by djakobczak1** [comment]
> These functions are only available in certain image versions: the ones available in this [repo](https://github.com/MariaDB/mariadb-docker). This also includes the current enterprise images, they have the same entrypoint.

I think currently in the `mdb.spec.image` description there is a note that only official images are supported so making this assumption does not sound as regression. However, if in the future you would like to support e.g. [bitnami mariadb images](https://github.com/bitnami/containers/tree/main/bitnami) (_this could be difficult_) then this approach may require deeper analysis.

To sum this up, we are interested in support of official mariadb images (10.6+), so from my side, both solutions should be fine,

**Comment by mmontes11** [comment]
Hey @djakobczak1 ! 

Initialization via a `Job` has been supported in the following PR: https://github.com/mariadb-operator/mariadb-operator/pull/423

This will be released in `v0.0.26`
